### ISSUE TITLE
[Feature] Upcasting of lamba method

### ISSUE THREAD

**Issue by johanhaleby** [jobrunr/jobrunr#40]
JobRunr looks great and I found it when searching for a "command scheduling library" that can be used together with [Occurrent](https://occurrent.org). 

The one drawback I can see with the library [this](https://www.jobrunr.io/en/documentation/background-methods/):

> Important: all your servers must run the same version of your code! If your webapp server has a newer version with a method signature that is not compatible with the server that processes your background jobs, a NoSuchMethod Exception will be thrown and job processing will fail!

This is of course understandable, but I'm lacking documentation on what's happening when a `NoSuchMethodException` is thrown. Will the method be retried forever in this case so that one has a chance to fix it and deploy a new version?

This brings me to the actual feature request. It would be great to be able to configure "upcasting" or conversion functions that can be applied to the JobRunr configuration. For example, if I have scheduled a job like this:

```java
BackgroundJob.enqueue(() -> SomeClass.something("Hello, world!"));
```

but I change `something` to take two strings before the job is executed, e.g. `something(String, String)`, I assume I will run into a `NoSuchMethodException`. Then it would be great to be able to configure upcasting functions like this:

```java
JobRunr.configure()
	.upcasting(document -> {
		if (document.isTargeting(SomeClass::something) && document.hasParameterCount(1)) {
			return document.addParameter("default for second param");
		} else {
			return document;

		}
	})
```

Where `document` is a (potentially immutable) class that is wrapping the serialized data:

```
{
  "lambdaType": "org.jobrunr.jobs.lambdas.JobLambda",
  "className": "org.something.Something",
  "methodName": "somehing",
  "jobParameters": [
    {
      "className": "java.lang.String",
      "object": "a test"
    }
  ]
}
```

WDYT?

**Comment by rdehuyss** [comment]
Hi @johanhaleby - thanks for the nice feedback!

Well to clarify - when a `NoSuchMethodException` or a `NoSuchClassException` is thrown, the job will immediately go to the failed state - it won't be retried automatically as the code is off-course not there.

The idea would be that you fix the error (e.g. deploy the correct code on all instances) or that you delete the job and requeue the correct one.

I think I understand your use-case: continuous deployment while there are still old jobs? Before I start working on this feature, is it an option to wait until all jobs are completed? It's easier to add a method that returns how many jobs are still enqueued than the upcasting of the lambda (just trying to be lazy here :smile: ). This could then be used as a guard/lock before deploying.

**Comment by johanhaleby** [comment]
> Well to clarify - when a `NoSuchMethodException` or a `NoSuchClassException` is thrown, the job will immediately go to the failed state - it won't be retried automatically as the code is off-course not there.
> 
> The idea would be that you fix the error (e.g. deploy the correct code on all instances) or that you delete the job and requeue the correct one.

Thanks for the clarification. Can you re-queue these types of messages from the API? Maybe one idea would be a setting to retry them _once_ each time the application is started? This way you have a chance to fix the problems in the code and redeploy and make this work again without too much hassle. 

> 
> I think I understand your use-case: continuous deployment while there are still old jobs?

Yes, continuous deployment is one issue (but the deployment doesn't necessarily have to be continuous :)). 

> Before I start working on this feature, is it an option to wait until all jobs are completed? It's easier to add a method that returns how many jobs are still enqueued than the upcasting of the lambda (just trying to be lazy here 😄 ). This could then be used as a guard/lock before deploying.

I'm not quite sure that I understand. Let's say you have enqueued 3 jobs that are due next week that will invoke method A. You then change method A and redeploy your app. Now you will run into a problem in a week's time that you may only notice, as a surprise, when the job fails with a `NoSuchMethodException`. How would "waiting until all jobs are completed" help here? :)

**Comment by rdehuyss** [comment]
Ok, I understand your use case - it's completely valid.

Another approach I'm thinking of: 
- on startup JobRunr checks if all the scheduled jobs can be executed or it logs an error
- I'm thinking of also showing a message in the Dashboard with a message that some scheduled jobs can not be run using the current code

I like the idea of the upcasting feature but I'm considering making this part of the Pro package (the reason: I also hope to earn some money using JobRunr and I think this is a valid feature for the Pro version). Currently, a lot of the users always tell me the free version is more than good enough so they don't go for a Pro subscription :smile: These kinds of features may help on that.

What do you think?

P.s.: nice work on Occurent!

**Comment by johanhaleby** [comment]
> on startup JobRunr checks if all the scheduled jobs can be executed or it logs an error

👍 

> I'm thinking of also showing a message in the Dashboard with a message that some scheduled jobs can not be run using the current code

Also good idea imho :) 

> I like the idea of the upcasting feature but I'm considering making this part of the Pro package (the reason: I also hope to earn some money using JobRunr and I think this is a valid feature for the Pro version). Currently, a lot of the users always tell me the free version is more than good enough so they don't go for a Pro subscription 😄  These kinds of features may help on that.

Makes sense. I hope you manage to make a living out of it, I think more open source projects should be funded :) 

That being said, I think there should be some way recover from these errors programmatically. For example, maybe automatic retry of failed jobs on startup due to `NoSuchMethodException`/`NoSuchClassException` could be provided in the free version?

> P.s.: nice work on Occurent!

Thanks :) My plan is to write a section in the docs regarding "scheduling"/deadline and refer to this library :) 

Need to run, my daughter is calling to come help her with something! :)

**Comment by rdehuyss** [comment]
> That being said, I think there should be some way recover from these errors programmatically. For example, maybe automatic retry of failed jobs on startup due to NoSuchMethodException/NoSuchClassException could be provided in the free version?

I think that should be possible. It will result in the following user flow:
- Code works and jobs are being processed
- New version is deployed with an incompatible job signature method
  - this will be logged
  - this will be visible in the dashboard
- User/Developer fixes the problem (by adding old code back/ ... )
- New version is deployed with a compatible job signature
  - All jobs in failed state with exception JobClass/JobMethodNotFoundException will be rerun

Is this what you have in mind?

Once I see how JobRunr integrates with Occurent, I'm more than happy to also add a section to the JobRunr website on how to integrate with it! Do not hesitate to create a feature request for that.

**Comment by johanhaleby** [comment]
> > That being said, I think there should be some way recover from these errors programmatically. For example, maybe automatic retry of failed jobs on startup due to NoSuchMethodException/NoSuchClassException could be provided in the free version?
> 
> I think that should be possible. It will result in the following user flow:
> 
> * Code works and jobs are being processed
> * New version is deployed with an incompatible job signature method
>   
>   * this will be logged
>   * this will be visible in the dashboard
> * User/Developer fixes the problem (by adding old code back/ ... )
> * New version is deployed with a compatible job signature
>   
>   * All jobs in failed state with exception JobClass/JobMethodNotFoundException will be rerun
> 
> Is this what you have in mind?

Exactly! :) A question regarding method dispatching though. If I change the method visibility from public to package private, will JobRunr still be able to execute the method? If so it would enable graceful upgrades, since you could then deprecate a the method (but retain it in the code so that JobRunr would still work). 

> 
> Once I see how JobRunr integrates with Occurent, I'm more than happy to also add a section to the JobRunr website on how to integrate with it! Do not hesitate to create a feature request for that.

Well since Occurrent is just a library, it should already integrate fine with JobRunr :). However, I was planning to write a small wrapper around JobRunr to make it a bit more "flexible". For example, I often use rich value objects in my domain model which are not serializable/deserializable using Jacksin. For example:

```java
public class Money {
	private final BigDecimal amount; 
	private final Currency currency; 

	private class Money(BigDecimal amount, Currency currency) {
		...
		Validate.greaterThanOrEqualToZero(amount);
	}

	Money add(Money money) {
		return ... 
	}

	..
}
```

From what I understand from the docs I won't be able to use such value objects with JobRunR?

```
BackgroundJob.enqueue(() -> SomeClass.something(Money.from(new BigDecimal("10"), SEK));
```

So what I was planning to do was to create a small utility to separate data from dispatch. I'm going to thinking out loud here but hopefully, it'll make sense :) What I mean by this is a utility, let's call it `AlarmClock` that you can use to schedule `Reminder`'s. Where `Reminder` is a simple data structure (POJO) that can be serializable/deserializable using Jackson. The `Reminder` could have this format: 

```
{ "type" : "reminder type", "version" : "2", "data" : ".." }
```

To use the "alarm clock" you do like this:

```java
SomePojo pojo = new SomePojo("hello", "world");
alarmClock.registerReminder(new Reminder(SomePojo.class.getSimpleName, "1", pojo));
```

What `registerReminder` will do is something like this:

```java
public void registerReminder(Reminder reminder) {
	BackgroundJob.enqueue(() -> reminderDispatcher.dispatch(reminder);
}
```

Here, `reminderDispatcher` would be a class that would never change (or at least it should be highly unlikely to change) so that there will be no backward incompatibilities (like the ones we've been talking about).

`ReminderDispatcher` would also be responsible for accepting "dispatch functions". I.e. on application starts up, the app should register functions that handles various reminders:

```java
reminderDispatcher.handle(reminder -> reminder.hasType(SomePojo.class.getSimpleName()), reminder -> SomeClass. something(reminder.data.getHello(), reminder.data.getWorld());
```

I.e. you register a `Predicate` and a `Consumer`/`Function` that should be invoked if the predicate matches the supplied reminder. This way you're free to refactor your application code (`SomeClass` and `something`) any way you find fit and still be able to handle the reminder.

As a nice benefit of this is that you'll also get a compile time error if you change `something` so that it no longer matches the lambda you suppled to `reminderDispatcher.handle(..)`, so you're forced to handle backward-incompatible changes. 

WDYT?

**Comment by rdehuyss** [comment]
Hi,

I had some busy times so did not get back immediately.

> Exactly! :) A question regarding method dispatching though. If I change the method visibility from public to package private, will JobRunr still be able to execute the method? If so it would enable graceful upgrades, since you could then deprecate the method (but retain it in the code so that JobRunr would still work).

This is already part of JobRunr Pro (again the fine line what to put in the FOSS version and the Pro version).

> From what I understand from the docs I won't be able to use such value objects with JobRunR?

This actually depends on the JSON Provider implementation you choose. E.g. if you annotate the static method in `Money` class with the correct Jackson annotation, all should be well.

> So what I was planning to do was to create a small utility to separate data from dispatch.
> ...
> WDYT

It looks ok - the only thing I'm always a little bit scared of is too many levels of abstraction. The reason I created JobRunr is that I like easy, readable, and concise code. I've seen many projects with too many levels of abstractions and that's something I try to avoid. In your case, I can see the added benefit though - It's always a difficult compromise.

**Comment by rdehuyss** [comment]
This has been resolved in JobRunr Pro v4.0.0. See the [documentation](https://www.jobrunr.io/en/documentation/pro/migrations/) and the [blog post](https://www.jobrunr.io/en/blog/2021-09-08-jobrunr-v4.0.0/) for more info.
